arwen/context/asyncCall.go

```diff
@@ -0,0 +1,286 @@
+package context
+
+import (
+	"math/big"
+
+	arwen "github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/ElrondNetwork/go-ext-wasm/wasmer"
+)
+
+func (host *vmContext) handleAsyncCallBreakpoint(result wasmer.Value, err error) (*vmcommon.VMOutput, error) {
+	host.SetRuntimeBreakpointValue(arwen.BreakpointNone)
+
+	convertedResult := arwen.ConvertReturnValue(result)
+	senderVMOutput := host.createVMOutput(convertedResult.Bytes())
+	intermediaryVMOutput := senderVMOutput
+
+	// If SC code is not found, it means this is either a cross-shard call or a wrong call.
+	dest := host.asyncCallDest
+	calledSCCode, err := host.GetCode(dest)
+	if err != nil || len(calledSCCode) == 0 {
+		// TODO detect same Shard call - this makes the empty calledSCCode an error
+		// if intraShard {
+		//   vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		//   return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError)
+		// }
+		return intermediaryVMOutput, nil
+	}
+
+	// Start calling the destination SC, synchronously.
+	destinationCallInput, err := host.createDestinationContractCallInput()
+	if err != nil {
+		vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError), nil
+	}
+
+	destinationVMOutput, err := host.executeOnNewContextAndGetVMOutput(destinationCallInput)
+	// TODO handle this error properly
+	// TODO pass error to the SC callback (append as argument)
+	// TODO consume remaining gas
+	if err != nil {
+		vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError), nil
+	}
+
+	intermediaryVMOutput = mergeTwoVMOutputs(intermediaryVMOutput, destinationVMOutput)
+
+	callbackCallInput, err := host.createCallbackContractCallInput(destinationVMOutput)
```

Comment:
Is is possible to extract logic to a function called `executeCallback()`?

Reply:
This section has been refactored.